core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java

```diff
@@ -126,10 +129,68 @@ protected DescriptorImpl() {
             Downloadable.all().add(createDownloadable());
         }
 
-        protected Downloadable createDownloadable() {
+        /**
+         * function that creates a {@link Downloadable}.
+         * @return a downloadable object
+         */
+        public Downloadable createDownloadable() {
+            if (this instanceof DownloadFromUrlInstaller.DescriptorImpl) {
+                final DownloadFromUrlInstaller.DescriptorImpl delegate = (DownloadFromUrlInstaller.DescriptorImpl)this;
+                return new Downloadable(getId()) {
+                    public JSONObject reduce(List<JSONObject> jsonList) {
+                        return delegate.reduce(jsonList);
+                    }
+                };
+            }
             return new Downloadable(getId());
         }
 
+        private JSONObject reduce(List<JSONObject> jsonList) {
+            List<ToolInstallerEntry> reducedToolEntries = new LinkedList<>();
+            //collect all tool installers objects from the multiple json objects
+            for (JSONObject jsonToolList : jsonList) {
+                ToolInstallerList toolInstallerList = (ToolInstallerList) JSONObject.toBean(jsonToolList, ToolInstallerList.class);
+                reducedToolEntries.addAll(Arrays.asList(toolInstallerList.list));
+            }
+
+            while (Downloadable.hasDuplicates(reducedToolEntries, "id")) {
```

Comment:
:bug: I feel the logic is very ineffective here. We already have N^2 in the method just returning a boolean value. And we process the list again in cycles below...